Multiple timezone support

arkokoley opened the pull request:
Fixes Issue #19 

## How this works?

Whenever a user visits the website, the user timezone is guessed and a cookie "browser.timezone" is set. For every subsequent request, that cookie is sent along and is used to set the Rails Time Zone for that request. 

Since the requests are threaded and Time.zone is thread safe, changing the timezone for one request doesn't affect other requests being processed at the same time.

arkokoley committed:
Implement timezone support for the user

Users will now get the webpage in their own timezone rather than being
served content in the server's timezone

Fix issue #19

arkokoley committed:
Tests for multiple-timezone-support

Fix issue #19

jdcc commented on app/controllers/admin/application_controller.rb:
```diff
@@ -27,6 +28,23 @@ def authorize_instance_access
   end
 
+  def set_time_zone
+    begin
```
What is this being-ensure doing for you?

jdcc commented on app/controllers/admin/application_controller.rb:
```diff
@@ -27,6 +28,23 @@ def authorize_instance_access
 
+  def set_time_zone
+    begin
+      logger.info 'I am the before action'
```
Try not to commit logging messages unless they're useful long-term. Can we pull these out or log the timezone change more informatively?

arkokoley replied:
The log lines were something I forgot to remove. I'll just remove them.

jdcc commented on app/controllers/application_controller.rb:
```diff
@@ -2,8 +2,25 @@ class ApplicationController < ActionController::Base
+  private
+
+    def set_time_zone
+      begin
```
What is this being-ensure doing for you?

arkokoley replied:
So the around filter wraps the entire request, unlike a [before or after] filter. Everything between `begin` and `yield` acts as a `before filter` and everything after that acts like an `after filter`.

jdcc replied:
I know. That's the function of `yield`, but the `ensure` isn't doing anything, so I'm not sure why you've got everything in a `begin...ensure` block.

jdcc commented on app/views/layouts/application.html.slim:
```diff
@@ -2,15 +2,17 @@ doctype html
     = stylesheet_link_tag  'application', media: 'all'
     = javascript_include_tag '//maps.google.com/maps/api/js?v=3.23&key=' + ENV["MAPS_API_KEY"]
     = javascript_include_tag '//cdn.rawgit.com/mahnunchik/markerclustererplus/master/dist/markerclusterer.min.js'
+    = javascript_include_tag '//cdnjs.cloudflare.com/ajax/libs/jstimezonedetect/1.0.6/jstz.min.js'
```
Can you download these and add them to the app, and then include them in the asset pipeline? We don't expect this app to need to scale such that a CDN is giving us much, and it's adding an external dependency we don't need. It's probably worth moving all three of these js libs into sprockets.

arkokoley replied:
I'll move these libs to the asset pipeline.

jdcc commented on spec/requests/timezones_spec.rb:
```diff
@@ -0,0 +1,24 @@
+    it "should correctly set user timezone" do
+      get root_url
+      assert_equal 200, status
+      expect(response.body).to include(cookies["browser.timezone"])
```
It looks like these tests are really just testing whether cookies work. Is there some element of the UI that reflects the timezone changes that we can test for?

jdcc commented on app/assets/javascripts/main.js:
```diff
@@ -30,4 +30,17 @@ $(function() {
+  window.BrowserTZone || (window.BrowserTZone = {});
+
+  BrowserTZone.setCookie = function() {
+    console.log(jstz.determine().name());
```
Do these console logs still need to be in here?

arkokoley replied:
No they don't. I'll remove these.

arkokoley commented on app/controllers/application_controller.rb:
```diff
@@ -2,8 +2,25 @@ class ApplicationController < ActionController::Base
+        old_time_zone = Time.zone
+        Time.zone = browser_timezone if browser_timezone.present?
+        yield
+        Time.zone = old_time_zone
```
In retrospect, it would be better to put this line in the `ensure` block to ensure predictable behavior.

arkokoley committed:
Show open time slots according to user timezone

Also removes extra logs and moves js from cdn to assets
pipeline

Implements #19

arkokoley committed:
Update tests for issue #19

arkokoley committed:
Merge branch 'dev' into multiple-timezone-support
